### ISSUE TITLE
V2 - Proposed Dev Fixes/Optimizations

### ISSUE THREAD

**Issue by MaximumCrash** [makerdao/community#890]
@twblack88  @amy-jung 

These are a list of outstanding issues with the site I believe can be allieviated to provide a better experience:

1. ~Try to get rid of the NavigationProvider. Currently it generates data needed by the Sidenav, Breadcrumbs, and Language Selector. While in practice this works, the problem lies in that the context provider gets updates "every single time we change a page". This can possibly be optimized by moving layout related components like the sidenav to site_layout, and leveraging the location data that comes from it. Then updating the Sidenav component to use a Static Query very similarly to the [hasura execution](https://github.com/hasura/gatsby-gitbook-starter/blob/master/src/components/sidebar/index.js). Be aware however that their example only works because they have 1 sidebar for their entire site. We can have X amount of sidebars depending on how many top section directories exist.~

2. ~Try to make a build process for auto generating breadcrumb data and language selector options during build time rather than on the fly. This is in relation to problem number 1.~

3. ~Figure out if there are conflicts with motion.divs and sx properties that are causing the flashing emotion styles on load. I don't think it is. But I've moved the layout and messed around in gatsby-browser and gatsby-ssr for days. Will need some time to pull the spaghetti apart. I'm trying to do that now with my tweaks to site-layout.~ (It's not entirely the motion divs fault. Some client side loading is causing things to take a long time to load.)

4. Fix the search results not scrolling with your arrow keys. This is a custom implementation so a little bit of math is required to check if the current element highlighted is out of view. 

5. ~Don't bold names in the header. They cause reflow.~

6. ~Fix the sidebar causing massive reflow when going from a landing page to a page with the sidenav.~ 

7. Improve the excerpt provided to lunr in gatsby-config. It's kind of basic and bare minimum, but it could be a bit cleaner.

8. ~Figure out how to speed up locale redirect on the index page. Currently we use a useEffect because we need the window to get access to a user's navigator locale. I've tried useLayoutEffect in the past, but it doesn't consistently push the user to the next page. I've also tried checking for undefined window and only running the navigate code on the client side, also doesn't consistently push the user to the next page.~

9. ~Improve the header file to not be so large. There's a lot of moving parts in there that can be decoupled (like the light/dark toggles), making a seperate mobile header component that handles just the scrolling logic, ect.~

10. ~Pull back on the automagic where you can. There's a lot of automatic generation happening which leads to longer build times and thus longer development times. If there's something that can be manually controlled, opt for it first.~

11. ~Create a gatsby-plugin for the all-contributors.json plugin. Currently there's a node script that converts the allcontributors file to json that's tacked on to the `yarn start` and `yarn build` commands. Would be better to have this built into the build process for the site.~ (Not doing, would be nice to have, but is out of scope for the needs of this project)

12. (Ongoing) ~Clean up styles to implement style variants instead of using sx everywhere. I focused on getting the project done, and we're here but the styles could leverrage the theme-ui's spec more so.~

**Comment by MaximumCrash** [comment]
@twblack88 Continued work recently:

8. I tested out more possibilities to execute this, and while I've considered createRedirect or strapping in some functionality to the 404 page, neither of these options would help in speeding this up. The only reason we can't do createRedirect is because redirecting is entirely based on a users window locale, which is not accessible during build time naturally. The 404 solution also wouldn't work, since it'd would be 2 redirects before a user get's to their intended homepage, not very cool. 

9. I cleaned up some of the code to decouple parts I found too strapped into the return structure, but there wasn't a good way to really "shrink" down this file, since in the context of mobile there's a lot of moving parts. I would love to have the mobile nav and header be one component but due to the mobile nav hosting the sidenav data, and that data specifically being loaded during build time in mdx_layout I think the execution we have now for our specific design case is as good as it can get for the time being. 

10. This was done by removing a number of config files that weren't being changed often enough for them be as accessible to content creators. Specifically social, footer, 404, and header mdx files have been removed and transplanted into their respective components with certain copy lines being moved to UI.json (making them accessible via the t function). My decision for this is that it doesn't make too much sense to oversupport the data control of these components and pages and that if there are changes happening to these components there are larger conversations happening around them, instead of a one off tweak. Removing these files also speeds up build time a tad as well. 

12. This will be a continously tweaked task as I've come to discover that the culprit of the flashing css style issue is not something we can directly influence with changes in how theme-ui is set up. The TLDR is that emotion.sh (the css-in-js library leveraged by theme-ui) applies it's styles using style tags and style tags are super unperformant. However, due to the complex rendering nature of these tags in tandem with their relative dom elements, if we have styles that are overtly complex there will be moments where they won't render fast enough to paint the window. This stackoverflow talks about the downsides of style tags: https://stackoverflow.com/questions/44371227/when-and-how-do-browsers-render-style-tag-in-body
